sunscreen_runtime/src/builder.rs

```diff
@@ -0,0 +1,648 @@
+            Self {
+                runtime,
+                statements: vec![],
+                messages: vec![],
+                witness: vec![],
+                compiled_zkp_program: None,
+                shared_inputs: vec![],
+                private_inputs: vec![],
+                public_inputs: vec![],
+                constant_inputs: vec![],
+            }
+        }
+
+        /// Encrypt a plaintext, adding the encryption statement to the proof.
+        ///
+        /// If you do not want to add the encryption statement to the proof, just use [the
+        /// runtime](`crate::GenericRuntime::encrypt`) directly.
+        pub fn encrypt<P>(&mut self, message: &P, public_key: &'k PublicKey) -> Result<Ciphertext>
+        where
+            P: TryIntoPlaintext + TypeName,
+        {
+            let pt = self.plaintext_typed(message)?;
+            self.encrypt_internal(Message::Plain(pt), public_key, None)
+        }
+
+        /// Encrypt a plaintext intended for sharing.
+        ///
+        /// The returned `SharedMessage` can be used:
+        /// 1. to add an encryption statement of ciphertext equality to the proof (see [`Self::encrypt_shared`]).
+        /// 2. as a shared input to a ZKP program (see [`Self::shared_input`]).
+        pub fn encrypt_and_share<P>(
+            &mut self,
+            message: &P,
+            public_key: &'k PublicKey,
+        ) -> Result<(Ciphertext, SharedMessage)>
+        where
+            P: ShareWithZkp + TryIntoPlaintext + TypeName,
+        {
+            // The user intends to share this message, so add a more conservative bound
+            let plaintext_typed = self.plaintext_typed(message)?;
+            let idx_start = self.messages.len();
+            let ct = self.encrypt_internal(
+                Message::Plain(plaintext_typed.clone()),
+                public_key,
+                Some(self.mk_bounds::<P>()),
+            )?;
+            let idx_end = self.messages.len();
+            // TODO shouldn't be assuming bulletproofs here...
+            // need to separate the notion of sharing duplicate messages and sharing to ZKP
+            let zkp_type = P::ZkpType::<BulletproofsFieldSpec>::type_name();
+            let shared_message = SharedMessage {
+                id: idx_start,
+                message: Arc::new(SharedPlaintextTyped {
+                    plaintext_typed,
+                    zkp_type,
+                }),
+                len: idx_end - idx_start,
+            };
+            Ok((ct, shared_message))
+        }
+
+        /// Encrypt a shared message, adding the new encryption statement to the proof.
+        ///
+        /// This method purposefully reveals that two ciphertexts enrypt the same underlying value. If
+        /// this is not what you want, use [`Self::encrypt`].
+        ///
+        /// This method assumes that you've created the `message` argument with _this_ builder.
+        pub fn encrypt_shared(
```

Comment:
`encrypt_and_share` and `encrypt_shared` are similar names, maybe we could have something slightly different?

Reply:
In the PR desc I talked about changing `share` to `link` but other than that, I struggle to think of better names. There's `encrypt_again` rather than `encrypt_shared` that seems even worse. `reencrypt`? :shrug: